keps/sig-node/4188-kubelet-pod-readiness-api/README.md

```diff
@@ -0,0 +1,451 @@
+components to get node-local Pod readiness information as this would not have
+any impact on other Nodes in the cluster.
+
+#### Kubelet restarts issue
+
+There is a Kubelet restart
+[issue](https://github.com/kubernetes/kubernetes/issues/100277) that could
+impact this API. After restart Kubelet marks all Pods as NotReady and then
+starts the init loop checking the actual pod's status. When Kubelet is in this
+`unknown` state it should not serve the API data which might be incorrect. This
+API needs to be reliable and returned data should reflect the actual state of
+the Pods. We don't want this API to return partial data. We should resolve this
+issue before the beta release. One of the possible fixes is returning error code
+indicating that Kubelet is in the init phase and that pod's data are not known
+yet.
+
+## Design Details
+
+### Proposed API
+
+We propose to add new gPRC API `status` in Kubelet, listening on a unix socket at `/var/lib/Kubelet/status/Kubelet.sock`. The endpoint will be versioned.
+The gRPC Service will expose 3 methods serving local Pods statuses data:
+
+```protobuf
+service Status {
+    // ListPodConditions returns a of List of PodConditions
+    rpc ListPodConditions(Empty) returns (stream PodConditionsResponse) {}
+    // WatchPodConditions returns a stream of List of PodConditions
+    // Whenever a pod state change api returns the new list
+    rpc WatchPodConditions(Empty) returns (stream PodConditionsResponse) {}
+    // GetPodConditions returns a PodInfo for given pod's UID
+    rpc GetPodConditions(PodConditionsRequest) returns (PodInfo) {}
+}
+
+// PodCondition aligns with v1.PodCondition.
+message PodCondition {
+    ConditionType Type = 1;
+    ConditionStatus Status = 2;
+    Timestamp LastProbeTime = 3;
+    Timestamp LastTransitionTime = 4;
+    string Reason = 5;
+    string Message = 6;
+}
+
+// ConditionType aligns with v1.PodConditionType
+enum ConditionType {
+    ContainersReady = 0;
+	Initialized = 1;
+	Ready = 2;
+	PodScheduled = 3;
+	DisruptionTarget = 4;
+}
+
+// ConditionStatus aligns with v1.ConditionStatus
+enum ConditionStatus {
+    True = 0;
+	False = 1;
+	Unknown = 2;
+}
+
+// PodInfo returns a Pod details and list of status Conditions with deletion info.
+message PodInfo {
+    string podUID = 1;
+    string podNamespace = 2;
+    string podName = 3;
+    bool static = 4;
+	repeated PodCondition conditions = 5;
+    Timestamp DeletionTimestamp = 3;
+}
+
+// PodConditionsResponse returns a stream of List of PodConditionsResponse
+// Whenever a Pod state changes it will return the new list.
+message PodConditionsListResponse {
+	repeated PodInfo Pods = 1;
+}
+
+// PodConditionsRequest contains Pods UID
+message PodConditionsRequest {
+    string podUID = 1;
+}
+```
+
+##### Unit tests
+
+
+##### e2e tests
+
+  - Feature gate enable/disable tests.
+  - Test API work when Kubelet restarts.
+  - Test API work when Control Plane is unreachable.
```

Comment:
let's also add a test for the kubelet in Standalone mode

Reply:
done